### PR TITLE
[iOS][documentation] Add Links to Related Examples

### LINKED ISSUE #13111: [iOS][documentation] Link relevant API classes to `/ios-sdk/maps/examples`
### Quick overview
Mapbox hosts example code for popular implementations of the iOS Maps SDK at www.mapbox.com/ios-sdk/maps/examples. To help interested users understand correct/recommended usage, these examples should be linked within the API docs, similar to the GL JS API's "Related" sections:

screen shot 2018-10-16 at 10 56 08 am

### Examples to be linked:
There are currently 39 live examples on https://www.mapbox.com/ios-sdk/maps/examples. I think it should be possible to link all of them to relevant parts of the documentation.

### PR DESCRIPTION
**Pull request by riastrad**
Closes #13111

This PR links examples that have been published on www.mapbox.com/ios-sdk/maps/examples to their related classes via the API documentation.

### CODE REVIEW

**Comment by riastrad** [comment]
Here's what this is looking like so far:
### [`-setCamera:animated:`](https://www.mapbox.com/ios-sdk/api/4.5.0/Classes/MGLMapView.html#/c:objc(cs)MGLMapView(im)setCamera:animated:)
screen shot 2018-10-31 at 10 38 25 am

How does this look for linking to published examples on www.mapbox.com/ios-sdk/maps/examples? I'm relying on the keyword `@see` to flag these links. I see it's used a few other places in the docs to flag related iOS methods/attributes ([MGLLocationManager.distanceFilter](https://www.mapbox.com/ios-sdk/api/4.5.0/Protocols/MGLLocationManager.html#/c:objc(pl)MGLLocationManager(py)distanceFilter), [.desiredAccuracy](https://www.mapbox.com/ios-sdk/api/4.5.0/Protocols/MGLLocationManager.html#/c:objc(pl)MGLLocationManager(py)desiredAccuracy), [.activityType](https://www.mapbox.com/ios-sdk/api/4.5.0/Protocols/MGLLocationManager.html#/c:objc(pl)MGLLocationManager(py)activityType), [a few places in MGLMapView](https://www.mapbox.com/ios-sdk/api/4.5.0/Classes/MGLMapView.html#/c:objc(cs)MGLMapView(py)preferredFramesPerSecond), etc.). I wanted to use an existing keyword to give some quick formatting, but if you're worried this might cause confusion, please just let me know and I can run with something else.

Section order is currently tied to our Jazzy theme templates, so if we want to change that we'll need to update the template files.

Any thoughts/concerns the direction of this update @mapbox/maps-ios?

**Comment by riastrad** [comment]
The approach of using `@see` does have the potential for creating too much visual clutter:
screen shot 2018-10-31 at 3 10 07 pm

Two thoughts on this: 
1. Might be worth only linking the top 4(ish) most relevant examples for any class/method
2. Maybe using some markdown to create a bulleted list (similar to GL JS) would be a cleaner way to approach this.

I'll focus for linking all examples up first, then do a quick review and clean-up. Should be too hard to tweak once all the proper links are in the right place.

**Comment by riastrad** [comment]
Toyed around with using a markdown unordered list instead of the `@see` keyword. Here are the results:

screen shot 2018-11-01 at 10 51 04 am

Personally, I prefer this formatting much less than the boxed visual callout that results from using `@see`. So I'm going to charge ahead with `@see` and limit example linking to four relevant examples.

**Comment by julianrex** [comment]
What do you think about making it clear that we're linking to an example, and have some minimal relevant description? Too much clutter?

e.g. instead of 
> `Camera animation`

we have something like:

> See the `Camera Animation` example to learn how to create a camera that rotates around the center point.

**Comment by riastrad** [comment]
@julianrex I'm definitely open to this since it would add more clarity to what the links are actually pointed toward and prevent any misguided/unnecessary link-clicking confusion. 

My only caveat is that I would probably end up limiting to 2 examples maximum to avoid clutter. But thinking on that a bit more it would probably better focus the examples I'm linking to - so I'm definitely up for it. 👍

**Comment by riastrad** [comment]
Based on team input, I've update the formatting to look like so: 
image

I'll be completing the example linking in this format going forward.

**Comment by riastrad** [comment]
I _think_ that all my iOS tests are failing because https://github.com/mapbox/mapbox-gl-native/pull/13184 is still in flight. I'll keep an eye on that and rebase once it's merged.

**Comment by riastrad** [comment]
@mapbox/maps-ios, can I get some insight into why these tests are failing for only some `platform/darwin` files? Seeing the error message:
> ✘ These generated Darwin files have modifications:

Doesn't look like rebasing to include the Xcode-10 circle updates did anything to resolve them.

**Comment by friedbunny** [comment]
Certain headers are generated — particularly those having to do with style layers. This means that, rather than edit the `.h` files, you edit the `.h.ejs` templates and then run [`make darwin-update-examples`](https://github.com/mapbox/mapbox-gl-native/blob/947af3059c5d2dec3dfdd8b8a6002679a6ccc055/platform/ios/DEVELOPING.md#adding-a-code-example) and/or `make darwin-style-code` to regenerate the actual source files.

File: platform/darwin/src/MGLLight.h [hunk 1]
```diff
@@ -58,6 +58,9 @@ NS_INLINE MGLSphericalPosition MGLSphericalPositionMake(CGFloat radial, CLLocati
 
 /**
  An `MGLLight` object represents the light source for extruded geometries in `MGLStyle`.
+
+#### Related examples
+See the <a href="https://www.mapbox.com/ios-sdk/maps/examples/light-example/">Adjust light of 3D buildings</a> to learn how to create and modify the light source for 3D geometries.
```
Comment (friedbunny) [inline platform/darwin/src/MGLLight.h]: Hrm, this particular entry in the generated header/docs doesn’t appear to auto-wrap... 🤔

File: platform/ios/src/MGLMapViewDelegate.h [hunk 2]
```diff
@@ -41,6 +41,9 @@ NS_ASSUME_NONNULL_BEGIN
  method returns `YES`, this camera becomes the map view’s camera.
  @return A Boolean value indicating whether the map view should stay at
  `oldCamera` or change to `newCamera`.
+ 
+ #### Related examples
+ See the <a href="https://www.mapbox.com/ios-sdk/maps/examples/constraining-gestures/">Restrict map panning to an area</a> example to learn how to use this method and `MGLMapCamera` objects to restrict a users ability to pan your map.
```
Comment (friedbunny) [inline platform/ios/src/MGLMapViewDelegate.h]: This (and some of the next few in `MGLMapViewDelegate.h` didn’t get the wrapping treatment. 🙇

Comment (riastrad) [reply platform/ios/src/MGLMapViewDelegate.h]: 😳😅

File: platform/darwin/src/MGLLight.h (hunk 1 above)
Comment (riastrad) [reply platform/darwin/src/MGLLight.h]: Yeah, not sure why that is... Seems like digging through [generate-style-code.js](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/darwin/scripts/generate-style-code.js) at this point is a bit tangential for this PR. I'm making a note to dig through it separately and I'll open up a new issue later this week.

File: platform/darwin/src/MGLCircleStyleLayer.h [hunk 3]
```diff
@@ -71,6 +71,16 @@ typedef NS_ENUM(NSUInteger, MGLCircleTranslationAnchor) {
  To display circles on the map whose radii correspond to real-world distances,
  use many-sided regular polygons and configure their appearance using an
  `MGLFillStyleLayer` object.
+ 
+ #### Related examples
```
Comment (1ec5) [inline platform/darwin/src/MGLCircleStyleLayer.h]: This header (and many others in this PR) are generated by `make darwin-style-code`, so much of the PR will be wiped away the next time someone runs that command. Make these changes in the .ejs files in platform/darwin/src/, then run `make darwin-style-code` to verify that they get generated as expected.

File: platform/darwin/src/MGLLight.h (hunk 1 above)
Comment (1ec5) [reply platform/darwin/src/MGLLight.h]: Call `wrap()` on `doc`:

https://github.com/mapbox/mapbox-gl-native/blob/35a0811eeb7b27378f1e37d942bfa4df2f44c6a8/platform/darwin/src/MGLLight.h.ejs#L66

like this:

https://github.com/mapbox/mapbox-gl-native/blob/35a0811eeb7b27378f1e37d942bfa4df2f44c6a8/platform/darwin/src/MGLLight.h.ejs#L74

**Comment by 1ec5** [comment]
Following up from https://github.com/mapbox/mapbox-gl-native/pull/13236#issuecomment-436002796, each property that has an example should have an `example` field in style-spec-overrides-v8.json. Then platform/darwin/src/MGLStyleLayer.h.ejs can refer to that `example` field after each call to `propertyDoc()`. For example, after:

https://github.com/mapbox/mapbox-gl-native/blob/5a70f2b15081bede66a2d736dc3e8f90ba73d601/platform/darwin/src/MGLStyleLayer.h.ejs#L120

add:

```ejs
<% if (property.example) { %>
 #### Related examples
 propertyExample(property).wrap(80, 1)
<% } -%>
```

Add a `propertyExample` function to style-spec-overrides-v8.json that returns the sentence based on `property.example`.

**Comment by riastrad** [comment]
@1ec5 for generated headers, I've successfully migrated the the inline comment logic to the `.ejs` templates and `generate-style-code.js`. 

If you could quickly confirm that this new implementation doesn't include any glaring errors, I would greatly appreciate it.

**Comment by friedbunny** [comment]
@riastrad That’ll be because of #13322 — if you rebase on current master to pull in #13327 here, all of the builds should pass again.

**Comment by riastrad** [comment]
Looks like code coverage slipped slightly compared to master, but this should be addressed once I rebase. Also seeing as this PR does not modify any code or functionality, it is independent of any test coverage concerns.

**Pull request merged**